### PR TITLE
Add generic pixel support

### PR DESCRIPTION
**Pull request by andreldsa**
#### What is the purpose of this pull request?
Pixel architecture MVP.

#### What problem is this solving?
Communication between the pixel apps and the store.

#### How should this be manually tested?
https://pixel--storecomponents.myvtex.com/samsung-s9-x/p

#### Types of changes
* [x] New feature (a non-breaking change which adds functionality)

### CODE REVIEW

File: react/PixelContext.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+const PixelContext = React.createContext()
+
+export const Pixel = (WrappedComponent) => {
+  return class Pixel extends Component {
+    constructor(props) {
+      super(props)
+    }
+
+    renderComponent = (context) => {
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: Remove the parentheses.

File: react/PixelContext.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
+import React, { Component } from 'react'
+
+const PixelContext = React.createContext()
+
+export const Pixel = (WrappedComponent) => {
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: Remove the parentheses.

File: react/PixelContext.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  return class Pixel extends Component {
+    constructor(props) {
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: Is this constructor necessary? I don't think so, because it's the default.

File: react/PixelContext.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    renderComponent = (context) => {
+      return (
+        <WrappedComponent subscribe={context.subscribe} context={context}/>
+      )
+    }
+
+    render() {
+      return (
+        <PixelContext.Consumer>
+          {(context) => this.renderComponent(context)}
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: Remove the parentheses.

File: react/PixelContext.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        </PixelContext.Consumer>
+      )
+    }
+  }
+}
+
+export const pixelGlobalContext = (WrappedComponent) => {
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: Remove the parentheses.

File: react/PixelContext.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  return class WithPixel extends Component {
+
+    constructor(props) {
+      super(props)
+      this.state = {
+        subscribers: []
+      }
+    }
+
+    notifySubscribers = (data) => {
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: Remove the parentheses.

File: react/PixelContext.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+      this.state.subscribers.forEach(subscriber => {
+        if (subscriber[data.event]) {
+          subscriber[data.event](data)
+        }
+      })
+    }
+    
+    push = (data) => {
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: Remove the parentheses.

File: react/PixelContext.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      }
+
+      if (this.state.subscribers.length === 0) {
+        setTimeout(notifyAndPush, 100)
+      } else {
+        notifyAndPush()
+      }
+    }
+    
+    subscribe = (subscriber) => {
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: Remove the parentheses.

File: react/PixelContext.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    }
+    
+    push = (data) => {
+      const notifyAndPush = () => {
+        this.notifySubscribers(data)
+        window.dataLayer.push(data)
+      }
+
+      if (this.state.subscribers.length === 0) {
+        setTimeout(notifyAndPush, 100)
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: I think that is better to create a constant for this "Magic number".

File: react/PixelContext.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        })
+      }
+    }
+
+    render() {
+      return (
+        <PixelContext.Provider value={{
+          subscribe: this.subscribe,
+          ...this.state
+        }}>
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: Put the ">" in a new line.

File: react/StoreContextProvider.js [hunk 11]
```diff
@@ -19,7 +20,8 @@ class StoreContextProvider extends Component {
   }
 
   pushToDataLayer = obj => {
-    window.dataLayer.push(obj)
+    console.log(">>> Push to datalayer")
```
Comment (vitoriaHeliane) [inline react/StoreContextProvider.js]: Remove the log.

File: react/PixelContext.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        if (subscriber[data.event]) {
+          subscriber[data.event](data)
+        }
+      })
+    }
+    
+    push = (data) => {
+      const notifyAndPush = () => {
+        this.notifySubscribers(data)
+        window.dataLayer.push(data)
```
Comment (vitoriaHeliane) [inline react/PixelContext.js]: I think you should verify if the ```window.dataLayer``` isn't undefined before access its function.

File: react/PixelContext.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
+import React, { Component } from 'react'
+
+const PixelContext = React.createContext()
+
+const SUBSCRIPTION_TIMEOUT = 100
+
+export function Pixel(WrappedComponent) {
```
Comment (brunojdo) [inline react/PixelContext.js]: Docs

File: react/PixelContext.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+        <PixelContext.Consumer>
+          {context => this.renderComponent(context)}
+        </PixelContext.Consumer>
+      )
+    }
+  }
+}
+
+export function pixelGlobalContext(WrappedComponent) {
+  return class WithPixel extends Component {
```
Comment (brunojdo) [inline react/PixelContext.js]: displayName

File: react/StoreContextProvider.js [hunk 15]
```diff
@@ -19,7 +20,7 @@ class StoreContextProvider extends Component {
   }
 
   pushToDataLayer = obj => {
```
Comment (brunojdo) [inline react/StoreContextProvider.js]: I think this has to be moved in Pixel component

File: react/PixelContext.js [hunk 16]
```diff
@@ -0,0 +1,103 @@
+import React, { Component } from 'react'
+import { withRuntimeContext } from 'render'
+
+const PixelContext = React.createContext()
+
+const SUBSCRIPTION_TIMEOUT = 100
+
+const GTM_UNDEFINED = ({ workspace, account }) => (
```
Comment (brunojdo) [inline react/PixelContext.js]: remove?

Comment (brunojdo) [inline react/PixelContext.js]: remove?

File: react/ProductContextProvider.js [hunk 17]
```diff
@@ -154,6 +154,17 @@ class ProductContextProvider extends Component {
       pageInfo,
       {
         event: 'productView',
+        products: [
```
Comment (brunojdo) [inline react/ProductContextProvider.js]: Is it an array?

File: CHANGELOG.md [hunk 18]
```diff
@@ -9,6 +9,10 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Added
 - Enables querying of `shippingData` in `orderFormQuery`
 - Adds `updateOrderFormShipping`, a mutation and function on `OrderFormContext`
+- `Pixel` context with store dispatch events.
+
+### Removed
+- `Google Tag Manager` script and it's logic.
```
Comment (vitoriaHeliane) [inline CHANGELOG.md]: its* :hammer:

File: react/PixelContext.tsx [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+  push: (data: PixelData) => void
+}
+
+const PixelContext = React.createContext<ContextType>({
+  push: () => undefined,
+  subscribe: () => () => undefined,
+})
+
+/**
+ * Pixel it's the HOC Component that provides an event subscription to the
```
Comment (vitoriaHeliane) [inline react/PixelContext.tsx]: Pixel is the ... * :hammer:

File: react/StoreContextProvider.js [hunk 20]
```diff
@@ -1,83 +1,64 @@
...
 const APP_LOCATOR = 'vtex.store'
 const CONTENT_TYPE = 'text/html;charset=utf-8'
 const META_ROBOTS = 'index, follow'
 
-const GTM_UNDEFINED = `No Google Tag Manager ID is defined. Take a look at:\
-https://${global.__RUNTIME__.workspace}--${global.__RUNTIME__.account}.myvtex.com/admin/apps/vtex.store/setup`
-
 class StoreContextProvider extends Component {
   static propTypes = {
+    runtime: PropTypes.shape({
```
Comment (vitoriaHeliane) [inline react/StoreContextProvider.js]: No JSDoc?

Comment (lucasecdb) [reply react/StoreContextProvider.js]: runtime is a common prop type

File: react/PixelContext.js [hunk 21]
```diff
@@ -76,11 +76,11 @@ export function pixelGlobalContext(WrappedComponent) {
     }
     
     subscribe = subscriber => {
-      console.log(">>>>>> subs: ", subscriber)
       if (subscriber) {
-        this.setState({
-          subscribers: [subscriber, ...this.state.subscribers]
-        })
+        this.setState(state => ({
+          ...state,
```
Comment (Klynger) [inline react/PixelContext.js]: I don't think you need to spread the state, this probably works without it.

File: react/StoreContextProvider.js [hunk 22]
```diff
@@ -47,37 +35,36 @@ class StoreContextProvider extends Component {
...
     } = settings
+
     const scripts = gtmId ? [{
       'type': 'application/javascript',
       'innerHTML': gtmScript(gtmId),
     }] : []
     const noscripts = gtmId ? [{ id: 'gtm_frame', innerHTML: gtmFrame(gtmId) }] : []
 
-    this.initDataLayer()
+    window.dataLayer = window.dataLayer || []
```
Comment (andreldsa) [inline react/StoreContextProvider.js]: Should use initDataLayer, once the structure needs to be empty every page view.

**Pull request merged**